core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheFileInStream.java

```diff
@@ -0,0 +1,202 @@
+    Objects.requireNonNull(b, "Read buffer cannot be null");
+    return read(ByteBuffer.wrap(b), off, len);
+  }
+
+  @Override
+  public int read(ByteBuffer byteBuffer, int off, int len) throws IOException {
+    Preconditions.checkArgument(off >= 0 && len >= 0 && len + off <= byteBuffer.capacity(),
+        PreconditionMessage.ERR_BUFFER_STATE.toString(), byteBuffer.capacity(), off, len);
+    Preconditions.checkState(!mClosed, "Cannot do operations on a closed BlockInStream");
+    if (len == 0) {
+      return 0;
+    }
+    if (mPos == mLength) {
+      return -1;
+    }
+    readChunk();
+    if (mCurrentChunk == null) {
+      closeDataReader();
+      if (mPos < mLength) {
+        throw new OutOfRangeException(String.format("Block %s is expected to be %s bytes, "
+                + "but only %s bytes are available in the UFS. "
+                + "Please retry the read and on the next access, "
+                + "Alluxio will sync with the UFS and fetch the updated file content.",
+             mLength, mPos));
+      }
+      return -1;
+    }
+    int toRead = Math.min(len, mCurrentChunk.readableBytes());
+    byteBuffer.position(off).limit(off + toRead);
+    mCurrentChunk.readBytes(byteBuffer);
+    mPos += toRead;
+    if (mPos == mLength) {
+      // a performance improvement introduced by https://github.com/Alluxio/alluxio/issues/14020
+      closeDataReader();
+    }
+    return toRead;
+  }
+
+  @Override
+  public int positionedRead(long position, byte[] buffer, int offset, int length)
+      throws IOException {
+    if (length == 0) {
+      return 0;
+    }
+    if (position < 0 || position >= mLength) {
+      return -1;
+    }
+
+    int totalBytesRead = 0;
+    try (DataReader reader = mGrpcReaderFactory.create(position, length)) {
+      while (totalBytesRead < length) {
+        DataBuffer dataBuffer = null;
+        try {
+          dataBuffer = reader.readChunk();
+          if (dataBuffer == null) {
+            break;
+          }
+          int bytesRead = dataBuffer.readableBytes();
+          dataBuffer.readBytes(buffer, offset, bytesRead);
+          totalBytesRead += bytesRead;
+          offset += bytesRead;
+        } finally {
+          if (dataBuffer != null) {
+            dataBuffer.release();
+          }
+        }
+      }
+    }
+    if (totalBytesRead == 0) {
+      return -1;
+    }
+    return totalBytesRead;
+  }
+
+  @Override
+  public long getPos() throws IOException {
+    return mPos;
+  }
+
+  @Override
+  public void seek(long pos) throws IOException {
+    Preconditions.checkState(!mClosed, "Cannot do operations on a closed BlockInStream");
+    Preconditions.checkArgument(pos >= 0, PreconditionMessage.ERR_SEEK_NEGATIVE.toString(), pos);
+    Preconditions.checkArgument(pos <= mLength,
+        "Seek position past the end of the read region (block or file).");
+    if (pos == mPos) {
+      return;
+    }
+    closeDataReader();
```

Comment:
if the new position is still within the range of the current chunk, we do not have to close the reader?

Reply:
hmmm, very likely you're right, but I just follow the code in the blockInStream.  I'm thinking if we could optimize this in another pr?

Reply:
sure not a blocker for this PR